typescript/core/src/ChunkCursor.ts

```diff
@@ -211,80 +160,75 @@ export class ChunkCursor {
-          continue;
-        }
-      }
-
-      this.#messageIndexCursors.push({
-        index: startIndex,
-        channelId: result.record.channelId,
-        records: result.record.records,
-      });
+      this.#orderedMessageOffset = this.#orderedMessageOffset.concat(result.record.records);
     }
 
     if (offset !== messageIndexesView.byteLength) {
       throw new Error(
         `${messageIndexesView.byteLength - offset} bytes remaining in message index section`,
       );
     }
-  }
 
-  #getSortTime(): bigint {
-    if (!this.#messageIndexCursors) {
-      return this.#reverse ? this.chunkIndex.messageEndTime : this.chunkIndex.messageStartTime;
+    if (this.#orderedMessageOffset.length === 0) {
+      return;
+    }
+
+    this.#orderedMessageOffset.sort(([logTimeA], [logTimeB]) => Number(logTimeA - logTimeB));
+    if (reverse) {
+      // If we used `logTimeB - logTimeA` as the comparator for reverse iteration, messages with
+      // the same timestamp would not be in reverse order. To avoid this problem we use reverse()
+      // instead.
+      this.#orderedMessageOffset.reverse();
+    }
+
+    const [logTimeFirstMessage] = this.#orderedMessageOffset[0]!;
+    if (logTimeFirstMessage < this.chunkIndex.messageStartTime) {
+      throw new Error(
+        `Chunk contains a message with logTime (${logTimeFirstMessage}) earlier than chunk messageStartTime (${this.chunkIndex.messageStartTime}) in chunk at offset ${this.chunkIndex.chunkStartOffset}`,
+      );
     }
 
-    const cursor = this.#messageIndexCursors.peek();
-    if (!cursor) {
+    const [logTimeLastMessage] = this.#orderedMessageOffset[this.#orderedMessageOffset.length - 1]!;
+    if (logTimeLastMessage > this.chunkIndex.messageEndTime) {
       throw new Error(
-        `Unexpected empty cursor for chunk at offset ${this.chunkIndex.chunkStartOffset}`,
+        `Chunk contains a message with logTime with logTime (${logTimeLastMessage}) later than chunk messageEndTime (${this.chunkIndex.messageEndTime}) in chunk at offset ${this.chunkIndex.chunkStartOffset}`,
       );
     }
 
-    return cursor.records[cursor.index]![0];
+    // Determine the indexes corresponding to the start and end time.
+    const startTime = reverse ? this.#endTime : this.#startTime;
+    const endTime = reverse ? this.#startTime : this.#endTime;
+    const iteratee = (logTime: bigint) => (reverse ? -logTime : logTime);
+    if (startTime != undefined) {
+      this.#nextMessageOffsetIndex = sortedIndexBy(this.#orderedMessageOffset, startTime, iteratee);
+    }
+    if (endTime != undefined) {
+      this.#messageOffsetEndIndex = sortedIndexBy(this.#orderedMessageOffset, endTime, iteratee);
+      // sortedIndexBy returns the minimum index but for the end index we actually want the highest index since
+      // endTime is inclusive. So we count up the end index manually until we reach a logTime that is not included anymore.
```

Comment:
It appears the lodash implementation (on which our sortedIndexBy is based) includes a flag for whether the highest value should be returned: https://github.com/lodash/lodash/blob/aa18212085c52fc106d075319637b8729e0f179f/src/.internal/baseSortedIndexBy.ts#L16

They use this to implement `sortedLastIndexBy`: https://github.com/lodash/lodash/blob/aa18212085c52fc106d075319637b8729e0f179f/src/sortedLastIndexBy.ts#L23

Seems like it might be slightly nicer (and slightly faster in cases when there are many messages with the same timestamp) to adopt this approach.

On the other hand it's probably not a very big impact.